File: sdk/go/subprocess_test.go
package shojiku

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"strings"
	"sync"
	"testing"
)

func assertEngineFailure(t *testing.T, err error, contains string) {
	t.Helper()
	if !errors.Is(err, ErrEngineFailure) {
		t.Fatalf("err = %v, want ErrEngineFailure", err)
	}
	if contains != "" && !strings.Contains(err.Error(), contains) {
		t.Errorf("message %q does not contain %q", err, contains)
	}
}

func assertUsage(t *testing.T, err error, contains string) {
	t.Helper()
	if !errors.Is(err, ErrUsage) {
		t.Fatalf("err = %v, want ErrUsage", err)
	}
	if contains != "" && !strings.Contains(err.Error(), contains) {
		t.Errorf("message %q does not contain %q", err, contains)
	}
}

// stubGenerate drives one render against a stub binary whose lifecycle
// behaviour is body.
func stubGenerate(t *testing.T, body string) (*Result, error) {
	t.Helper()
	return stubClient(t, stubBinary(t, body)).
		Generate(context.Background(), "receipt", map[string]any{})
}

func TestANonZeroExitWithNoReportIsAHostFailure(t *testing.T) {
	// Neither of the contract's two levels: nobody determined anything about
	// the document, so manufacturing a document failure would be a lie a
	// Success() check could swallow.
	_, err := stubGenerate(t, `echo "it broke" >&2; exit 3`)

	assertEngineFailure(t, err, "wrote no report")
	if !strings.Contains(err.Error(), "it broke") {
		t.Errorf("stderr is not quoted for a caller with nothing else to go on: %s", err)
	}
}

func TestAReportThatIsNotJsonIsAHostFailure(t *testing.T) {
	_, err := stubGenerate(t, `printf 'not json at all' > "$report"; exit 0`)

	assertEngineFailure(t, err, "is not JSON")
}

func TestJsonThatIsNotTheEnvelopeIsAHostFailure(t *testing.T) {
	_, err := stubGenerate(t, `printf '{"something":true}' > "$report"; exit 0`)

	assertEngineFailure(t, err, "not a report envelope")
}

func TestValidJsonWithTheWrongTypeIsNotReportedAsBrokenJson(t *testing.T) {
	// It parsed; it is simply not this envelope. Telling a caller "not JSON"
	// about valid JSON sends them looking in the wrong place.
	_, err := stubGenerate(t, `printf '{"ok":"yes"}' > "$report"; exit 0`)

	assertEngineFailure(t, err, "not a report envelope")
	if strings.Contains(err.Error(), "not JSON") {
		t.Errorf("valid JSON was reported as broken JSON: %s", err)
	}
}

func TestAReportPastTheReadCapIsRefusedRatherThanLoaded(t *testing.T) {
	_, err := stubGenerate(t,
		`{ printf '{"ok":true,"pad":"'; head -c 9000000 /dev/zero | tr '\0' 'a'; printf '"}'; } > "$report"; exit 0`)

	assertEngineFailure(t, err, "past this package's")
}

func TestAReportAtTheCapIsStillRead(t *testing.T) {
	// The reason one byte past the cap is read: a report AT the cap has to be
	// distinguishable from one past it.
	padding := strings.Repeat("a", maxReportBytes-len(`{"ok":true,"pad":""}`))

	rep, err := parseReport([]byte(`{"ok":true,"pad":"`+padding+`"}`), "")

	if err != nil {
		t.Fatalf("a report at the cap was refused: %v", err)
	}
	if !rep.ok {
		t.Error("the report at the cap did not read as ok")
	}
}

func TestABinaryThatWritesNothingAtAllIsAHostFailure(t *testing.T) {
	_, err := stubGenerate(t, `exit 0`)

	assertEngineFailure(t, err, "wrote no report")
}

func TestAnEngineWithoutTheReportCapabilityIsRefusedByName(t *testing.T) {
	path := stubBinaryWith(t, `exit 0`, `{"version":"0.0.1","capabilities":["text"]}`)

	_, err := stubClient(t, path).Generate(context.Background(), "receipt", map[string]any{})

	if !errors.Is(err, ErrIncompatibleEngine) {
		t.Fatalf("err = %v, want ErrIncompatibleEngine", err)
	}
	if !strings.Contains(err.Error(), reportCapability) {
		t.Errorf("the refusal does not name the capability key: %s", err)
	}
}

func TestACapabilityPayloadThatIsNotJsonIsAHostFailure(t *testing.T) {
	path := stubBinaryWith(t, `exit 0`, `{oops`)

	_, err := stubClient(t, path).EngineInfo(context.Background())

	assertEngineFailure(t, err, "not JSON")
}

func TestACapabilityPayloadThatIsNotAnObjectIsAHostFailure(t *testing.T) {
	for _, payload := range []string{`[1,2,3]`, `null`, `"a string"`} {
		t.Run(payload, func(t *testing.T) {
			path := stubBinaryWith(t, `exit 0`, payload)

			_, err := stubClient(t, path).EngineInfo(context.Background())

			assertEngineFailure(t, err, "not an object")
		})
	}
}

func TestAFailingCapabilitiesCallIsAHostFailure(t *testing.T) {
	// The probe's own process failing is a third thing again: the binary ran
	// and refused to say what it is.
	client := stubClient(t, rawStub(t, `echo "no such subcommand" >&2; exit 2`))

	_, err := client.EngineInfo(context.Background())

	assertEngineFailure(t, err, "exited 2")
	if !strings.Contains(err.Error(), "no such subcommand") {
		t.Errorf("stderr is not quoted: %s", err)
	}
}

func TestStderrOnASuccessfulRunIsIgnoredRatherThanParsed(t *testing.T) {
	// Prose is not a contract. A successful run that warned on stderr is a
	// success carrying whatever the REPORT said, and nothing else.
	result, err := stubGenerate(t,
		`echo "shojiku: warning[x] something" >&2; printf '{"ok":true,"diagnostics":{"items":[]}}' > "$report"; exit 0`)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Success() {
		t.Fatalf("the result failed: %v", result.Failure())
	}
	if len(result.Diagnostics()) != 0 {
		t.Errorf("stderr was parsed into %d diagnostics", len(result.Diagnostics()))
	}
}

func TestAnArgumentCarryingShellMetacharactersIsPassedThroughLiterally(t *testing.T) {
	// os/exec takes a slice and execs the binary directly: no shell runs, so
	// a value carrying $(…), a semicolon or a quote is ONE argument holding
	// those characters and nothing interprets it.
	hostile := `$(touch /tmp/pwned); rm -rf /; "'`
	argvLog := filepath.Join(t.TempDir(), "argv")
	client := stubClient(t, stubBinary(t, fmt.Sprintf(
		`for arg in "$@"; do printf '%%s\n' "$arg" >> %q; done
printf '{"ok":true,"diagnostics":{"items":[]}}' > "$report"
exit 0`, argvLog)))

	result, err := client.GenerateSource(context.Background(),
		Source{Template: sourceTemplate(textItem("who")), AssetsDir: hostile},
		map[string]any{"who": "x"})

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Success() {
		t.Fatalf("the call failed: %v", result.Failure())
	}
	logged, err := os.ReadFile(argvLog)
	if err != nil {
		t.Fatalf("reading the recorded argv: %v", err)
	}
	if !slices.Contains(strings.Split(string(logged), "\n"), hostile) {
		t.Errorf("the hostile value did not cross as ONE argument; child saw:\n%s", logged)
	}
	if _, err := os.Stat("/tmp/pwned"); err == nil {
		t.Error("a shell interpreted the argument")
	}
}

func TestAnEngineInfoCallCrossesTheSameWayAsALifecycleCall(t *testing.T) {
	info, err := newTestClient(t).EngineInfo(context.Background())

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if info["version"] == nil {
		t.Error("the payload carries no version")
	}
	if !hasCapability(info, reportCapability) {
		t.Errorf("the real engine does not advertise %q", reportCapability)
	}
}

func TestAProcessThatCannotBeStartedAtAllIsAHostFailure(t *testing.T) {
	// A running system cannot produce this on demand — the binary was proved
	// to be an executable file when the client was built — so the transport
	// is driven directly with a path that has since stopped being one.
	client := stubClient(t, stubBinary(t, `exit 0`))
	client.settings.engine.binary = &binary{path: t.TempDir(), source: "configuration"}
	client.settings.engine.probed = true

	_, _, _, err := client.settings.engine.spawn(context.Background(), []string{"render"}, nil)

	assertEngineFailure(t, err, "could not run")
}

func TestTheCapabilityProbeRunsOncePerEngine(t *testing.T) {
	// Asking a binary what it can do is a whole process; asking twice per
	// render would double the cost of every call.
	counter := stubBinary(t, `printf '{"ok":true,"diagnostics":{"items":[]}}' > "$report"; exit 0`)
	client := stubClient(t, counter)

	for range 3 {
		if _, err := client.Generate(context.Background(), "receipt", map[string]any{}); err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
	}

	if !client.settings.engine.probed {
		t.Error("the engine was never marked as probed")
	}
}

func TestTheProbeStaysOncePerEngineUnderConcurrentCallers(t *testing.T) {
	// The clause the SDKs in languages without threads got for free. A Go
	// client is documented as safe for concurrent use, so "once" has to mean
	// once even when several goroutines arrive together.
	client := stubClient(t, stubBinary(t,
		`printf '{"ok":true,"diagnostics":{"items":[]}}' > "$report"; exit 0`))

	var wg sync.WaitGroup
	for range 8 {
		wg.Add(1)
		go func() {
			defer wg.Done()
			_ = client.settings.engine.requireReport(context.Background())
		}()
	}
	wg.Wait()

	if !client.settings.engine.probed {
		t.Error("the engine was never marked as probed")
	}
}

func TestAFailingProbeIsRetriedRatherThanRemembered(t *testing.T) {
	// Only SUCCESS is remembered: a binary that could not answer once has not
	// been established to be incapable forever.
	client := stubClient(t, stubBinaryWith(t, `exit 0`, `{oops`))

	first := client.settings.engine.requireReport(context.Background())
	second := client.settings.engine.requireReport(context.Background())

	assertEngineFailure(t, first, "")
	assertEngineFailure(t, second, "")
	if client.settings.engine.probed {
		t.Error("a failed probe was remembered as a success")
	}
}

func TestAPerCallVariableReachesTheChildWithoutJoiningTheClientsEnvironment(t *testing.T) {
	base := []string{"HOME=/root"}

	composed := childEnv(base, map[string]string{"SHOJIKU_PASSPHRASE": "secret"})

	if len(composed) != 2 || composed[1] != "SHOJIKU_PASSPHRASE=secret" {
		t.Errorf("childEnv = %v, want the per-call variable appended", composed)
	}
	if len(base) != 1 {
		t.Error("the client's own environment was mutated")
	}
	if got := childEnv(base, nil); len(got) != 1 {
		t.Errorf("childEnv with nothing extra = %v, want the base unchanged", got)
	}
}

File: sdk/php/tests/SubprocessTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Engine;
use Shojiku\Exception\EngineFailureException;
use Shojiku\Exception\IncompatibleEngineException;
use Shojiku\Report;

/**
 * The transport's own failure modes.
 *
 * The contract has two failure levels — caller error, and something a
 * document did — and both arrive in the report. None of these is either: the
 * process died, wrote nothing, or wrote something that is not the envelope.
 * Manufacturing a document failure out of that would tell the caller
 * something about their document that nobody determined.
 */
final class SubprocessTest extends TestCase
{
    use EngineFixtures;
    use StubBinary;

    public function testANonZeroExitWithNoReportIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            $stub = $this->stubBinary($dir, 'echo "something went very wrong" >&2'."\n".'exit 3');

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/wrote no report.*something went very wrong/');
            $this->client(['binary' => $stub])->generate('receipt', []);
        });
    }

    public function testAReportThatIsNotJsonIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            $stub = $this->stubBinary($dir, 'printf "not json at all" > "$report"');

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/is not JSON/');
            $this->client(['binary' => $stub])->generate('receipt', []);
        });
    }

    public function testJsonThatIsNotTheEnvelopeIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            // Valid JSON, wrong shape: `ok` is what every branch downstream
            // reads, so a payload without it is not a report.
            $stub = $this->stubBinary($dir, 'printf \'{"ok":"yes"}\' > "$report"');

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/not a report envelope/');
            $this->client(['binary' => $stub])->generate('receipt', []);
        });
    }

    public function testAReportPastTheReadCapIsRefusedRatherThanLoaded(): void
    {
        $this->inTempDir(function (string $dir): void {
            $stub = $this->stubBinary($dir, sprintf(
                'head -c %d /dev/zero | tr "\\0" "x" > "$report"',
                Report::MAX_BYTES + 1,
            ));

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/past this package/');
            $this->client(['binary' => $stub])->generate('receipt', []);
        });
    }

    public function testABinaryThatWritesNothingAtAllIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            $stub = $this->stubBinary($dir, 'exit 0');

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/wrote no report$/');
            $this->client(['binary' => $stub])->generate('receipt', []);
        });
    }

    public function testAnEngineWithoutTheReportCapabilityIsRefusedByName(): void
    {
        $this->inTempDir(function (string $dir): void {
            $old = '{"version":"0.0.1","capabilities":["render.pdf"],"builtinLocales":["en-US"]}';
            $stub = $this->stubBinary($dir, 'exit 0', $old);

            $this->expectException(IncompatibleEngineException::class);
            $this->expectExceptionMessageMatches('/does not advertise `cli.report`/');
            $this->client(['binary' => $stub])->generate('receipt', []);
        });
    }

    public function testACapabilityPayloadThatIsNotJsonIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            $stub = $this->stubBinary($dir, 'exit 0', 'not json');

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/capability payload is not JSON/');
            $this->client(['binary' => $stub])->engineInfo();
        });
    }

    public function testACapabilityPayloadThatIsNotAnObjectIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            $stub = $this->stubBinary($dir, 'exit 0', '"just a string"');

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/capability payload is not an object/');
            $this->client(['binary' => $stub])->engineInfo();
        });
    }

    public function testAFailingCapabilitiesCallIsAHostFailure(): void
    {
        $this->inTempDir(function (string $dir): void {
            $path = $dir.'/broken';
            file_put_contents($path, "#!/bin/sh\necho 'no such subcommand' >&2\nexit 2\n");
            chmod($path, 0o755);

            $this->expectException(EngineFailureException::class);
            $this->expectExceptionMessageMatches('/exited 2.*no such subcommand/');
            $this->client(['binary' => $path])->engineInfo();
        });
    }

    public function testStderrOnASuccessfulRunIsIgnoredRatherThanParsed(): void
    {
        // The engine prints its diagnostics to stderr as prose as well as
        // into the report. Prose is not a contract, so nothing here reads it:
        // a run whose stderr is noise still succeeds, and the diagnostics the
        // caller gets are the report's.
        $result = $this->client()->generate('warns', []);

        self::assertTrue($result->success());
        self::assertCount(1, $result->warnings());
    }

    public function testAnArgumentCarryingShellMetacharactersIsPassedThroughLiterally(): void
    {
        // `proc_open` gets an ARRAY, so no shell is involved and there is
        // nothing to quote for. A name that would be a command substitution
        // in a shell is refused as a template NAME — by the name rules, not
        // by anything having executed it.
        $marker = sys_get_temp_dir().'/shojiku-php-injection-'.bin2hex(random_bytes(6));
        $result = $this->client()->generate('$(touch '.$marker.')', []);

        self::assertTrue($result->failed());
        self::assertFileDoesNotExist($marker);
    }

    public function testAnEngineInfoCallCrossesTheSameWayAsALifecycleCall(): void
    {
        $info = $this->client()->engineInfo();

        self::assertArrayHasKey('version', $info);
    }

    public function testAProcessThatCannotBeStartedAtAllIsAHostFailure(): void
    {
        // {@see Binary} has already proved the path is an executable file, so
        // a running system cannot produce this — which is exactly why the
        // guard is exercised directly rather than left unproven.
        $this->expectException(EngineFailureException::class);
        $this->expectExceptionMessageMatches('/could not run `\/bin\/nothing`/');
        Engine::started(false, '/bin/nothing');
    }

    public function testAStartedProcessIsHandedBackUntouched(): void
    {
        $handle = fopen('php://memory', 'r+');
        self::assertNotFalse($handle);

        self::assertSame($handle, Engine::started($handle, '/bin/shojiku'));
        fclose($handle);
    }

    public function testTheCapabilityProbeRunsOncePerEngine(): void
    {
        // The check a linked SDK spends on an ABI revision costs a whole
        // process here, so it is asked once and remembered — two renders
        // through one client is one probe.
        $this->inTempDir(function (string $dir): void {
            $counter = $dir.'/probes';
            $stub = $this->stubBinary($dir, 'printf \'{"ok":true,"diagnostics":{"items":[]}}\' > "$report"');
            $script = (string) file_get_contents($stub);
            $script = str_replace(
                'if [ "$1" = "capabilities" ]; then',
                'if [ "$1" = "capabilities" ]; then'."\n"
                    .'    echo probe >> '.escapeshellarg($counter),
                $script,
            );
            file_put_contents($stub, $script);

            $client = $this->client(['binary' => $stub]);
            $client->generate('receipt', []);
            $client->generate('receipt', []);

            self::assertSame(1, count(file($counter) ?: []));
        });
    }
}
